src/System.Memory/src/System/Buffers/ReadOnlyBuffer_helpers.cs

```diff
@@ -0,0 +1,305 @@
+                case BufferType.Array:
+                    if (start.Segment != end.Segment)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    }
+
+                    if (endIndex - startIndex >= count)
+                    {
+                        return new SequencePosition(start.Segment, startIndex + count);
+                    }
+
+                    ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+                    return default;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal SequencePosition Seek(SequencePosition start, SequencePosition end, long count, bool checkEndReachable = true)
+        {
+            var startIndex = start.Index;
+            var endIndex = end.Index;
+            var type = GetBufferType();
+
+            startIndex = GetIndex(startIndex);
+            endIndex = GetIndex(endIndex);
+
+            switch (type)
+            {
+                case BufferType.MemoryList:
+                    if (start.Segment == end.Segment && endIndex - startIndex >= count)
+                    {
+                        // end.Index >= bytes + Index and end.Index is int
+                        return new SequencePosition(start.Segment, startIndex + (int)count);
+                    }
+                    return SeekMultiSegment((IMemoryList<byte>) start.Segment, startIndex, (IMemoryList<byte>) end.Segment, endIndex, count, checkEndReachable);
+
+
+                case BufferType.OwnedMemory:
+                case BufferType.Array:
+                    if (endIndex - startIndex >= count)
+                    {
+                        // end.Index >= bytes + Index and end.Index is int
+                        return new SequencePosition(start.Segment, startIndex + (int)count);
+                    }
+
+                    ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+                    return default;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static SequencePosition SeekMultiSegment(IMemoryList<byte> start, int startIndex, IMemoryList<byte> end, int endPosition, long count, bool checkEndReachable)
+        {
+            SequencePosition result = default;
+            var foundResult = false;
+            var current = start;
+            var currentIndex = startIndex;
+
+            while (current != null)
+            {
+                // We need to loop up until the end to make sure start and end are connected
+                // if end is not trusted
```

Comment:
What does it mean by `end is not trusted`?

Reply:
Position with end segment was passed in as an argument and can be not connected to the linked list current `ROB` has.